### ISSUE TITLE
Volume plugins fail when selinux is enabled

### ISSUE THREAD

**Issue by jfindley** [docker/docker#18005]
When a docker volume plugin is use, and docker daemon is running with --selinux-enabled, attempts to use a volume plugin will result in a confusing error message: ```Error response from daemon: no such file or directory```.  This appears to be because of selinux support, running the container with --security-opt label:disable fixes this.

Looking deeper into this, I believe this is related to [syscall_linux:138](https://github.com/docker/docker/blob/729c9a97822ebee2c978a322d37060454af6bc66/vendor/src/github.com/syndtr/gocapability/capability/syscall_linux.go#L138) where the daemon tries to set an selinux context on the volume path.  However, when using a volume plugin, path will evaluate to an empty string, hence the error message.

Strace output demonstrating this:
```
[pid  7442] read(61, "HTTP/1.1 200 OK\r\nContent-Type: appplication/vnd.docker.plugins.v1+json\r\nDate: Mon, 16 Nov 2015 09:16:08 GMT\r\nContent-Length: 3\r\n\r\n{}\n", 4096) = 133
[pid  7442] read(61, 0xc2085ae000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
[pid  7442] write(61, "POST /VolumeDriver.Path HTTP/1.1\r\nHost: /run/docker/plugins/rexray.sock\r\nUser-Agent: Go 1.1 package http\r\nContent-Length: 27\r\nAccept: application/vnd.docker.plugins.v1+json\r\n\r\n{\"Name\":\"rexray-test-vol\"}\n", 203) = 203
[pid  7442] epoll_wait(4, {}, 128, 0)   = 0
[pid  7149] <... sched_yield resumed> ) = 0
[pid  7442] epoll_wait(4,  <unfinished ...>
[pid  7421] <... read resumed> "time=\"2015-11-16T09:16:08Z\" level=info msg=\"getting path to volume\" driverName=docker volumeID= volumeName=rexray-test-vol \n", 32768) = 124
[pid  7442] <... epoll_wait resumed> {{EPOLLOUT, {u32=2667585840, u64=140031486337328}}}, 128, -1) = 1
[pid  7421] read(39,  <unfinished ...>
[pid  7442] epoll_wait(4,  <unfinished ...>
[pid  7149] select(0, NULL, NULL, NULL, {0, 20}) = 0 (Timeout)
[pid  7149] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  7211] write(11, "{\"log\":\"time=\\\"2015-11-16T09:16:08Z\\\" level=info msg=\\\"getting path to volume\\\" driverName=docker volumeID= volumeName=rexray-test-vol \\n\",\"stream\":\"stdout\",\"time\":\"2015-11-16T09:16:08.437652525Z\"}\n", 198) = 198
[pid  7149] <... select resumed> )      = 0 (Timeout)
[pid  7149] clock[pid  7149] clock[pid  7149] clock[pid  7149] clock[pid  d  7149] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  7211] epoll_wait(4, {}, 128, 0)   = 0
[pid  7149] <... select resumed> )      = 0 (Timeout)
[pid  7442] <... epoll_wait resumed> {{EPOLLIN|EPOLLOUT, {u32=2667585840, u64=140031486337328}}}, 128, -1) = 1
[pid  7149] sched_yield( <unfinished ...>
[pid  7442] read(61, "HTTP/1.1 200 OK\r\nContent-Type: appplication/vnd.docker.plugins.v1+json\r\nDate: Mon, 16 Nov 2015 09:16:08 GMT\r\nContent-Length: 19\r\n\r\n{\"Mountpoint\": \"\"}\n", 4096) = 150
[pid  7442] lstat("", 0xc2085a6090)     = -1 ENOENT (No such file or directory)
[pid  7442] lsetxattr("", "security.selinux", "system_u:object_r:svirt_sandbox_file_t:s0", 41, 0) = -1 ENOENT (No such file or directory)
[pid  7442] write(2, "time=\"2015-11-16T09:16:08.755380971Z\" level=error msg=\"Handler for POST /containers/create returned error: no such file or directory\" \n", 135) = 135
[pid  7149] <... sched_yield resumed> ) = 0
[pid  7442] write(2, "time=\"2015-11-16T09:16:08.757670389Z\" level=error msg=\"HTTP Error\" err=\"no such file or directory\" statusCode=404 \n", 115 <unfinished ...>
```

**Comment by jfindley** [comment]
[fedora@ip-172-22-2-189 ~]$ sudo docker version
Client:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version: docker-1.8.2-1.gitf1db8f2.fc22.x86_64
 Go version:   go1.4.2
 Git commit:   f1db8f2/1.8.2
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version:
 Go version:   go1.4.2
 Git commit:   f1db8f2/1.8.2
 Built:
 OS/Arch:      linux/amd64
[fedora@ip-172-22-2-189 ~]$ sudo docker info
Containers: 36
Images: 18
Storage Driver: devicemapper
 Pool Name: docker-253:0-8473056-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: xfs
 Data file: /dev/atomicos/docker-data
 Metadata file: /dev/atomicos/docker-meta
 Data Space Used: 517.7 MB
 Data Space Total: 104.7 GB
 Data Space Available: 104.2 GB
 Metadata Space Used: 3.211 MB
 Metadata Space Total: 218.1 MB
 Metadata Space Available: 214.9 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Library Version: 1.02.93 (2015-01-30)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.10-200.fc22.x86_64
Operating System: Fedora 22 (Twenty Two)
CPUs: 1
Total Memory: 1.953 GiB
Name: ip-172-22-2-189.eu-west-1.compute.internal
ID: HH4J:3CWD:HV4Z:F3N5:ZLFT:XNIW:A3LX:TU74:KYBX:Q57Z:HYNY:XKDB
Debug mode (server): true
File Descriptors: 20
Goroutines: 32
System Time: 2015-11-16T09:49:25.19535492Z
EventsListeners: 0
Init SHA1: 603c4f89aedd044c26f9b65e8f6af985190a4da3
Init Path: /usr/libexec/docker/dockerinit
Docker Root Dir: /var/lib/docker
[fedora@ip-172-22-2-189 ~]$ uname -a
Linux ip-172-22-2-189.eu-west-1.compute.internal 4.1.10-200.fc22.x86_64 #1 SMP Mon Oct 5 14:22:49 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by rhatdan** [comment]
Looks like we are trying to set the label on newly created file sysstems even if they do not support Xattrs.

In this case you should probably disable SELinux support for this container

```
docker run --security-opt=label:disable ...
```

**Comment by rhatdan** [comment]
```
 docker run --volume-driver rexray --volume rexray-test-vol:/data -ti busybox sh
```
Since rexray (Whatever that is) does not support XAttrs SELinux will likely block access to this volume anyways.

**Comment by rhatdan** [comment]
docker daemon should either only setxattr on vfs volumes. since it would not know if other volumes support xattrs.

We should look into whether or not rexray supports context mounts. If we could pass in the context="system_u:object_r:svirt_sandbox_file_t:s0:c1,c2" to the mount command.

**Comment by crosbymichael** [comment]
This is more of an issue with the specific plugin and not docker.  You should probably open the issue on the plugin repo to get help on what is supported.

**Comment by jfindley** [comment]
Hi, this is actually a docker problem not a plugin issue.

The strace output above is taken from the docker daemon, and the important line is:
```
lsetxattr("", "security.selinux", "system_u:object_r:svirt_sandbox_file_t:s0", 41, 0) = -1 ENOENT (No such file or directory)
```
This will always fail when using any volume plugin.  As pointed out above though, instead of fixing the path (which is probably impossible, as there may not actually be a path), docker needs to skip (or warn and continue on failure) the setxattr call when a volume plugin is in use.

**Comment by clintonskitson** [comment]
@rhatdan Thank you for the research here.  I suppose there are two outstanding solutions to the plugin issue re SELinux based on your research

1) Have plugin mount with option of "system_u:object_r:svirt_sandbox_file_t:s0:c1,c2"
  a) I am assuming docker would need to pass this information to the plugin as part of the `/mount` API call
2) Docker skip setxattr call with plugins
 a) Is this less secure?  Seems like a simple fix that would handle it for all plugins.

**Comment by rhatdan** [comment]
Yes docker should not attempt the setxattr if a path does not exist.  Also perhaps it should check the error code from this call to see if it is ENOSUP, since certain file systems will not support setting XAttrs, and just warn on this situation.

**Comment by thaJeztah** [comment]
ping @cpuguy83 could you have a look at this, and see if modifications have to be made?

**Comment by rhatdan** [comment]
@jfindley could you check if https://github.com/docker/docker/pull/19921 fixes your issue?

**Comment by jfindley** [comment]
@rhatdan Sorry for the delayed response, was out of the office for a few days.  It looks like the linked changeset only handless ENOTSUP, but the problem I have is different - the volume path is "" at this point, so I don't think it will help?  I'm happy to test if it'd be useful though.

**Comment by rhatdan** [comment]
@jfindley the SetFileLabel should ignore file paths set to "", I don't believe this is where the error is being generated, have you located the failing call in go?  Or is this where it fails?

**Comment by HackToday** [comment]
any progress for this security issue?  most user cares about why volume plugin not work with selinux.

**Comment by rhatdan** [comment]
What volume plugin are you using to cause this?  Can you give me a couple of line of code to recreate the issue, and I will look into it.

**Comment by HackToday** [comment]
@clintonskitson  could you give link for that issue for this ?

https://github.com/emccode/rexray/issues/180

This issue is still open for that.

And @rhatdan which patch fix the the issue #18005 it is still opening, not closed

**Comment by HackToday** [comment]
any new progress here ? Do we have anyone working on this?

**Comment by clintonskitson** [comment]
@rhatdan I was able to verify that with CoreOS, specifying the parameter you mentioned prior does make it work correctly.

`docker run -ti --security-opt=label:disable --volume-driver=rexray -v testing221:/tst busybox`

**Comment by HackToday** [comment]
hi @clintonskitson  and @rhatdan 

We would like to see, selinux can work with docker, and docker run containers not need to add extra options.

@rhatdan Could you confirm if selinux can work with docker in atomic and coreos ?  Thanks

**Comment by rhatdan** [comment]
I have no way to test rexray.  So I can not confirm.   We are building our own volume plugin and we do not see these calls being made.  I am not sure who is making the call, although in our case we would like the call to be made, but check the ENOSUP error and the allow it to continue.

**Comment by clintonskitson** [comment]
Follow Docker thread https://github.com/docker/docker/pull/20829#issuecomment-190852173

**Comment by HackToday** [comment]
Seems this pull is the newer 

Docker pull https://github.com/docker/docker/pull/20834

@clintonskitson

**Comment by jfindley** [comment]
Apologies for the lack of response - I've been out of the country lots, and unfortunately due to workload it's going to be a few weeks before I have time to test this.  My test case was based on fedora atomic, with a basic rexray config running in a container.  #20829 looks really promising, though.

**Comment by rhatdan** [comment]
Actually could you check out https://github.com/docker/docker/pull/20834 Which I believe is the correct fix for this issue.  Currently we are attempting to set a default label on all volumes, but the original intention was to set a label on only "local" volumes.

**Comment by HackToday** [comment]
@jfindley if you have fedora atomic environment, is it possible for you to try fix #20834 mentioned above by @rhatdan . Also it would be helpful if you provide your atomic fedora version, like `docker info` Which can help me figure out if my atomic could apply such fix and related steps to apply that, Thus I could verify on my environment. ( I used fedora 21 atomic)
